doc/man/man9/zaptel.9.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man9/zaptel.9.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH ZAPTEL 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Kernel Inerfaces"
.SH NAME
.B zaptel
\- Zapata Telephony functions
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <zapata.h>
.ad l
.HP
.BI "void zt_alarm_notify(struct zt_span *span);
.PD 0
.HP
.BI "struct zt_tone *zt_dtmf_tone(char " digit ", int " mf );
.HP
.BI "void zt_dynamic_receive(struct zt_span *" span ", unsigned char *" msg ", int " msglen );
.HP
.BI "int zt_dynamic_register(struct zt_dynamic_driver *" driver );
.HP
.BI "void zt_dynamic_unregister(struct zt_dynamic_driver *" driver );
.HP
.BI "void zt_ec_chunk(struct zt_chan *" chan ", unsigned char *" rxchunk ", const unsigned char *" txchunk );
.HP
.BI "void zt_ec_span(struct zt_span *" span );
.HP
.BI "void zt_hooksig(struct zt_chan *" chan ", zt_rxsig_t " rxsig );
.HP
.BI "void zt_init_tone_state(struct zt_tone_state *" ts ", struct zt_tone *" zt );
.HP
.BI "char *zt_lboname(int " lbo );
.HP
.BI "void zt_qevent_lock(struct zt_chan *" chan ", int " event );
.HP
.BI "void zt_qevent_nolock(struct zt_chan *" chan ", int " event );
.HP
.BI "void zt_rbsbits(struct zt_chan *" chan ", int " bits );
.HP
.BI "int zt_receive(struct zt_span *" span );
.HP
.BI "int zt_register(struct zt_span *" span ", int " prefmaster );
.HP
.BI "void zt_set_dynamic_ioctl(int (*" func ") (unsigned int " cmd ", unsigned long " data ));
.HP
.BI "short zt_tone_nextsample(struct zt_tone_state *" ts ", struct zt_tone *" zt );
.HP
.BI "int zt_transmit(struct zt_span *" span );
.HP
.BI "short zt_txtone_nextsample(struct zt_chan *" ss );
.HP
.BI "int zt_unregister(struct zt_span *" span );
.ad b
.PD
.\"
.\"
.SH ARGUMENTS
.PP
Various.
.\"
.\"
.SH OVERVIEW
.PP
The
.BR zaptel (4)
driver provides the Zapata Telephony interface as a
.I STREAMS
driver that provides the listed exported kernel functions.
.\"
.\"
.SH DESCRIPTION
.PP
These functions are currently undocumented.
.PP
Functions provided by the
.BR zaptel (4)
driver are as folows:
.PP
.TS
tab(:);
lb
l.
ZAPTEL Function
zaptel
zt_alarm_notify
zt_dtmf_tone
zt_dynamic_receive
zt_dynamic_register
zt_dynamic_unregister
zt_ec_chunk
zt_ec_span
zt_hooksig
zt_init_tone_state
zt_lboname
zt_qevent_lock
zt_qevent_nolock
zt_rbsbits
zt_receive
zt_register
zt_set_dynamic_ioctl
zt_tone_nextsample
zt_transmit
zt_txtone_nextsample
zt_unregister
.TE
.\"
.\"
.SH CONTEXT
.PP
Various.
.\"
.\"
.SH MP-STREAMS
.PP
Undocumented.
.\"
.\"
.SH NOTICES
.PP
The
.BR zaptel (4)
driver is planned, but currently incomplete.
.\"
.\"
.SH "SEE ALSO"
.PP
.IR STREAMS (9).
.\"
.\"
.SH BUGS
.PP
As the driver is incomplete, all bugs are unwritten.
.\"
.\"
.SH COMPATIBILITY
.PP
Zapata Telephony.
.\"
.\"
.SH CONFORMANCE
.PP
None.
.\"
.\"
.SH HISTORY
.PP
The
.BR zaptel (4)
driver is planned but incomplete.
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/modules/zaptel.c
/*****************************************************************************

 @(#) File: src/modules/zaptel.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/modules/zaptel.c (" PACKAGE_ENVR ") " PACKAGE_DATE;


/*
 *  This module pushes over an OpenSS7 MX Stream and provides a zaptel drvier
 *  interface to such a Stream.  This permits the approach of making the
 *  /dev/zap device an MX multiplexer device and simply autopushing this zaptel
 *  module over the MX multiplexer device Stream using autopush.  It will then
 *  be tranparent to the zaptel application (e.g. OpenPBX, Asterisk) as to
 *  whether the device opened was a Streams driver or the archaic character
 *  device driver that zaptel previously provided.
 *
 *  The other approoach is to simply provide a direct zaptel interface to the MX
 *  multiplexer.
 *
 *  Note that this module does not allow zaptel devices to be registered with
 *  the driver as that is the purpose of a driver.  There could be a zaptel-mx
 *  driver that allows zaptel devices to register with the zaptel-mx driver and
 *  then provides an MX interface that can then be linked under the MX
 *  multiplexer, or the MX multiplexer could be provided with the capability of
 *  registering zaptel device drivers.
 *
 *  So, this module is a bit of an experiment.
 */
#include <sys/os7/compat.h>

#include <ss7/lmi.h>
#include <ss7/lmi_ioctl.h>
#include <ss7/mxi.h>
#include <ss7/mxi_ioctl.h>
#include <ss7/zap.h>
#include <ss7/zap_ioctl.h>

#define ZAP_MX_DESCRIP	"ZAPTEL Multiplex (ZAP-MX) STREAMS Module"
#define ZAP_MX_EXTRA	"Part of the OpenSS7 VoIP Stack for Linux Fast-STREAMS"
#define ZAP_MX_REVISION	"OpenSS7 src/modules/zaptel.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define ZAP_MX_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define ZAP_MX_DEVICE	"Part of the OpenSS7 Stack for Linux Fast-STREAMS"
#define ZAP_MX_CONTACT	"Brian Biduock <bidulock@openss7.org>"
#define ZAP_MX_LICENSE	"GPL"
#define ZAP_MX_BANNER	ZAP_MX_DESCRIP		"\n" \
			ZAP_MX_EXTRA		"\n" \
			ZAP_MX_REVISION		"\n" \
			ZAP_MX_COPYRIGHT	"\n" \
			ZAP_MX_DEVICE		"\n" \
			ZAP_MX_CONTACT		"\n"
#define ZAP_MX_SPLASH	ZAP_MX_DESCRIP		" - " \
			ZAP_MX_REVISION		"\n"

#ifdef LINUX
MODULE_AUTHOR(ZAP_MX_CONTACT);
MODULE_DESCRIPTION(ZAP_MX_DESCRIP);
MODULE_SUPPORTED_DEVICE(ZAP_MX_DEVICE);
#ifdef MODULE_LICENSE
MODULE_LICENSE(ZAP_MX_LICENSE);
#endif				/* MODULE_LICENSE */
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-zap_mx");
#endif				/* MODULE_ALIAS */
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif
#endif				/* LINUX */

#define ZAP_MX_MOD_ID		CONFIG_STREAMS_ZAP_MX_MODID
#define ZAP_MX_MOD_NAME		CONFIG_STREAMS_ZAP_MX_NAME

/*
 *  =========================================================================
 *
 *  STREAMS Definitions
 *
 *  =========================================================================
 */

#define MOD_ID		ZAP_MX_MOD_ID
#define MOD_NAME	ZAP_MX_MOD_NAME
#ifdef MODULE
#define MOD_BANNER	ZAP_MX_BANNER
#else				/* MODULE */
#define MOD_BANNER	ZAP_MX_SPLASH
#endif				/* MODULE */

static struct module_info zap_minfo = {
	.mi_idnum = MOD_ID,		/* Module ID number */
	.mi_idname = MOD_NAME,		/* Module ID name */
	.mi_minpsz = 1,			/* Min packet size accepted */
	.mi_maxpsz = INFPSZ,		/* Max packet size accepted */
	.mi_hiwat = 1,			/* Hi water mark */
	.mi_lowat = 0,			/* Lo water mark */
};

STATIC streamscall int zap_open(queue_t *, dev_t *, int, int, cred_t *);
STATIC streamscall int zap_close(queue_t *, int, cred_t *);

STATIC struct qinit zap_rinit = {
	.qi_putp = ss7_oput,		/* Read put (message from below) */
	.qi_qopen = zap_open,		/* Each open */
	.qi_qclose = zap_close,		/* Last close */
	.qi_minfo = &zap_minfo,		/* Information */
};

STATIC struct qinit zap_winit = {
	.qi_putp = ss7_iput,		/* Write put (message from above) */
	.qi_minfo = &zap_minfo,		/* Information */
};

STATIC struct streamtab zap_mxinfo = {
	.st_rdinit = &zap_rinit,	/* Upper read queue */
	.st_wrinit = &zap_winit,	/* Upper write queue */
};

/*
 *  =========================================================================
 *
 *  Private Data Structures
 *
 *  =========================================================================
 */

/*
 * zaptel structure
 */
typedef struct zap {
	STR_DECLARATION (struct zap);	/* stream declaration */
	struct zap_config config;	/* configuration */
	struct zap_stats statsp;	/* stats periods */
	struct zap_stats stats;		/* statistics */
	struct zap_notify notify;	/* notifications */
} zap_t;

#define ZAP_PRIV(__q) ((struct zap *)(__q)->q_ptr)

struct zap_config zap_default = {
	.block_size = 64,		/* 64 bits per block */
	.encoding = MX_ENCODING_NONE,	/* sample block encoding */
	.sample_size = 8,		/* sample size */
	.rate = 8000,			/* clock rate */
	.tx_channels = 1,		/* tx channels */
	.rx_channels = 1,		/* rx channels */
	.opt_flags = MX_PARM_OPT_CLRCH,	/* option flags */
};

STATIC struct zap *zap_opens = NULL;

STATIC struct zap *zap_alloc_priv(queue_t *, struct zap **, dev_t *, cred_t *);
STATIC struct zap *zap_get(struct zap *);
STATIC void zap_put(struct zap *);
STATIC void zap_free_priv(queue_t *);

/*
 *  =========================================================================
 *
 *  PRIMITIVES
 *
 *  =========================================================================
 */
/*
 *  -------------------------------------------------------------------------
 *
 *  Primitives sent upstream
 *
 *  -------------------------------------------------------------------------
 *  Because we a emulating a legacy character device, we do not send primitives
 *  upstream.
 */

/**
 *  m_error: - send an M_ERROR message upstream if necessary
 *  @q: a queue in the stream
 *  @zap: stream private structure
 *  @error: error number (if any)
 */
STATIC int
m_error(queue_t *q, struct zap *zap, int error)
{
	mblk_t *mp;
	int hangup = 0;

	if (error < 0)
		error = -error;
	switch (error) {
	case EBUSY:
	case ENOBUFS:
	case ENOMEM:
	case EAGAIN:
		return (-error);
	case EPIPE:
	case ENETDOWN:
	case EHOSTUNREACH:
		hangup = 1;
	}
	if ((mp = ss7_allocb(q, 2, BPRI_MED))) {
		if (hangup) {
			mp->b_datap->db_type = M_HANGUP;
			zap->state = MXS_UNUSABLE;
			printd(("%s: %p: <- M_HANGUP\n", MOD_NAME, zap));
			putnext(zap->oq, mp);
			return (-error);
		} else {
			mp->b_datap->db_type = M_ERROR;
			*mp->bwptr++ = (error < 0) ? -error : error;
			*mp->bwptr++ = (error < 0) ? -error : error;
			zap->state = MXS_UNUSABLE;
			printd(("%s: %p: <- M_ERROR\n", MOD_NAME, zap));
			putnext(zap->oq, mp);
			return (QR_DONE);
		}
	}
	rare();
	return (-ENOBUFS);
};

/*
 *  -------------------------------------------------------------------------
 *
 *  Primitive sent downstream
 *
 *  -------------------------------------------------------------------------
 *  Because we a emulating a legacy character device we do not send primitives
 *  downstream.
 */

/*
 *  -------------------------------------------------------------------------
 *
 *  Primitives from below
 *
 *  -------------------------------------------------------------------------
 */
